### ISSUE TITLE
Building packages into the jupyterlite kernel

### ISSUE THREAD

**Issue by psychemedia** [r-wasm/jupyterlite-webr-kernel#7]
I'm exploring the possibility of using JupyeterLite / jupyterlite-webr-kernel in a secure (prison) environment. This requires having a self-contained jupyterlite environment running on a secured server.

Loading packages into webR jupyterlite notebooks seems to involve downloading the wasm compiled packages from a public repository on demand. Is there a way of bundling the packages into a custom built JupyterLite webr kernel?

**Comment by psychemedia** [comment]
Related to this, I wonder: is there a way of mounting a custom filesystem built with the https://github.com/r-wasm/actions/tree/main/.github/workflows#r-wasmactions-reusable-workflows Github Action with the JupyterLite kernel?

I would like to be able to ship a JupyterLite environment with custom packages bundled into it.

**Comment by georgestagg** [comment]
Yes, in principle this should be possible. The idea would be to build an R repository of the packages you'd like using [{rwasm}](https://github.com/r-wasm/rwasm) and then including this in your jupyterlite static output, so that the packages are served in the same way as jupyterlite. Probably the Wasm package repository can just be added to jupyterlite's files in a subdirectory.

You can serve either a package repository or a filesystem image, but I'd personally recommend a repo.

Then, once you have started your kernel under jupyterlite and are loading packages you can set the `repos=` argument to point to your repository relative to the served jupyterlite directory:

```r
install.packages("cli", repos = "./repo")
```

or, mount your filesystem image and setup `.libPaths()` to load packages from the mount location:
```r
webr::mount(
  mountpoint = "/packages",
  source = "./some/path/to/image/packages.data"
)
.libPaths(c("/packages", .libPaths()))
```

---

Ideally, we'd be able to set the default webR package repository from the jupyterlite configuration, avoiding the need for the `repos =` argument. This is not possible currently, but something that would be good to have. I'll leave this issue open for tracking that.

**Comment by psychemedia** [comment]
Thanks — will give that a try...

**Comment by georgestagg** [comment]
Its also possible to mount each package one-by-one, if you have difficulty building a combined `packages.data` file using the `rwasm` package. Perhaps it would be best to attempt that first - then move onto a combined VFS image once that is confirmed working.

See me comment on https://github.com/r-wasm/webr/issues/465#issuecomment-2312031244 for more info on that. The comment is targeting running webR on iOS, but the same idea should apply with Jupyterlite: serving the package data in some subdirectory, then using `webr:mount()` to make it availabe on the virtual filesystem.

**Comment by psychemedia** [comment]
I can get the packages to load from a relative path (not sure if there is a jupyterlite settling that also needs fettling?) but if I put the full repo url (eg `repos = "https://ouseful-demos.github.io/jupyterlite-m348-demo/notebooks/repo/"`) then that seems to work.

I wonder, can webR introspect and return the juoyterlite installation URL?
